daemon/mgr/spec.go

```diff
@@ -17,79 +20,41 @@ type SpecWrapper struct {
 	argsArr [][]string
 }
 
-// SetupFunc defines spec setup function type.
-type SetupFunc func(ctx context.Context, m *ContainerMeta, s *SpecWrapper) error
-
-var setupFunc = []SetupFunc{
-	// process
-	setupProcessArgs,
-	setupProcessCwd,
-	setupProcessEnv,
-	setupProcessTTY,
-	setupProcessUser,
-	setupCap,
-	setupNoNewPrivileges,
-	setupOOMScoreAdj,
-
-	// cgroup
-	setupCgroupCPUShare,
-	setupCgroupCPUSet,
-	setupCgroupCPUPeriod,
-	setupCgroupCPUQuota,
-	setupCgroupMemory,
-	setupCgroupMemorySwap,
-	setupCgroupMemorySwappiness,
-	setupDisableOOMKill,
-
-	// namespaces
-	setupUserNamespace,
-	setupNetworkNamespace,
-	setupIpcNamespace,
-	setupPidNamespace,
-	setupUtsNamespace,
-
-	// volume spec
-	setupMounts,
-
-	// network spec
-	setupNetwork,
-
-	// host device spec
-	setupDevices,
-
-	// linux-platform-specifc spec
-	setupSysctl,
-	setupAppArmor,
-	setupCapabilities,
-	setupSeccomp,
-	setupSELinux,
+func createSpec(ctx context.Context, c *ContainerMeta, specWrapper *SpecWrapper) error {
+	// new a default spec from containerd.
+	s, err := ctrd.NewDefaultSpec(ctx, c.ID)
+	if err != nil {
```

Comment:
First I want to say,  fields in specs have related with each other,  you can not focus on just one but not care about other.
For eg. When you set  namespace, you must be careful about you are use other container ns or use host , or just make a new ns . Because in different case, lots of spec field should be changed. If you set uts ns to host, you need modify hostname, if you set user ns, you should get correct values for `s.Linux.UIDMappings` and `s.Linux.GIDMappings`.
How can you say you can just care one part. Or you can say it can be fixed by set `setup functions` to a good order, but I think it bad. When we read code at first time, we just think they
are  general functions, these functions do not care about order since they are independent from each other.  Unfortunately, they do have order, and if they execute in the wrong order, things will go wrong way.

On the other hand, spec code is splited into different parts by `specs.Spec`.
```
type Spec struct {
    // Version of the Open Container Runtime Specification with which the bundle complies.
    Version string `json:"ociVersion"`
    // Platform specifies the configuration's target platform.
    Platform Platform `json:"platform"`
    // Process configures the container process.
    Process Process `json:"process"`
    // Root configures the container's root filesystem.
    Root Root `json:"root"`
    // Hostname configures the container's hostname.
    Hostname string `json:"hostname,omitempty"`
    // Mounts configures additional mounts (on top of Root).
    Mounts []Mount `json:"mounts,omitempty"`
    // Hooks configures callbacks for container lifecycle events.
    Hooks Hooks `json:"hooks"`
    // Annotations contains arbitrary metadata for the container.
    Annotations map[string]string `json:"annotations,omitempty"`

    // Linux is platform specific configuration for Linux based containers.
    Linux *Linux `json:"linux,omitempty" platform:"linux"`
    // Solaris is platform specific configuration for Solaris containers.
    Solaris *Solaris `json:"solaris,omitempty" platform:"solaris"`
    // Windows is platform specific configuration for Windows based containers, including Hyper-V containers.
    Windows *Windows `json:"windows,omitempty" platform:"windows"`
}
```
If you want to change one part, you can modify the corresponding files, but you still need to care about that if the modification will influence other fields. As I said before, spec fields are not  Independent from each other. That why I say the design before is not reasonable.
Maybe I am not express my meaning totally, we can continue to discuss, @rudyfly

Reply:
To be honest, I am not on the side of the original way. This refactoring is my dish. For me, I cannot accept to couple specWrapper with several managers with spec. It breaks the simplicity of code. Although it seems to be quite scalable, but redundant so much. 

We can achieve scalability by several ways. If ideal scalability is not included in this part, we can improve this.